### DISCUSSION TITLE
over_time with make_before_break is not taking effect during re-auth if the remote certificate is either revoked or ocsp responder is not reachable

### DISCUSSION THREAD

**Discussion by pk-wc** [strongswan/strongswan#2349] [Q&A]
### Purpose
To validate cert revoked scenario by enabling make_before_break, reauth_time = 600s and revocation = ifuri.

### Expectation
New tunnel should fail to come-up while the existing tunnel remains up for 60s (over_time = 10% of reauth_time) before it goes down.

### What is being observed?
Scenario 1 - If cert is revoked then both the tunnels (existing and new) are going down immediately on receiving the cert revoked response.
Scenario 2 - If cert is valid but ocsp responder is not reachable then the tunnels remain up for 10 seconds until curl request times-out and then the tunnels go down.

### Configurations
Following are the configurations of swanctl.conf and strongswan.conf,

**swanctl.conf**
```
connections {
    home {
        dpd_delay = 20s
        local_addrs = <masked>
        proposals = aes256-sha1-prfsha1-modp2048
        reauth_time = 600s
        remote_addrs = <masked>
        version = 2
        vips = ::
        children {
            home {
                esp_proposals = aes256-sha1
                life_time = 10h
                remote_ts = ::0/0
                updown = /usr/lib/strongswan/ipsec/_updown iptables
            }
        }
        local {
            auth = pubkey
            certs = <masked>
            id = <masked>
        }
        remote {
            auth = pubkey
            id = %any
            revocation = ifuri
        }
    }
}
```

**strongswan.conf**
```
charon {
        threads = 16

        close_ike_on_child_failure = yes
        group = config-ipsec
        retransmit_tries = 5
        retransmit_timeout = 8
        retransmit_base = 1
        install_routes = no
        replay_window = 0
        keep_alive = 20s

        make_before_break = yes

        plugins {

                revocation {
                        enable_ocsp = yes
                        enable_crl = no
                }
                error-notify {
                }
                sql {
                        loglevel = -1
                }
                updown{
                      dns_handler=yes
                }
                systime-fix{
                       threshold = 1971
                       threshold_format = %Y
                       interval = 180
                       reauth = yes
                }
        }
}

pluto {

}

libstrongswan {

}
```

Is there a configuration that we are missing out over here ?
Please let us know.

Thanks,
Prasad

**Comment by tobiasbrunner** [comment]
Please provide logs (see #196).

**Reply by pk-wc** [reply]
Attaching the logs and configurations.
In these logs there are 2 connections home and home2. Revocation test was being performed only on home tunnel.
[cert_revoked_with_make_before_break_logs_m.txt](https://github.com/user-attachments/files/16237181/cert_revoked_with_make_before_break_logs_m.txt)
[swanctl_conf.txt](https://github.com/user-attachments/files/16237182/swanctl_conf.txt)

**Reply by tobiasbrunner** [reply]
OK, I see what happens. With make-before-break reauthentication, we delay the online certificate verification, which usually happens during the IKE_AUTH exchange. That's because if the OCSP/CRL check is done via VPN (if it's up), it might not work while creating the replacement SA. The reason for that is the order in which initiator and responder install the CHILD_SA during IKE_AUTH (the responder has the SA already installed when the initiator receives the last response and it tries to verify the certificate but has not yet installed the SA on its end). That's also why traffic is interrupted during a reauthentication (as compared to rekeyings where that's not the case), although with make-before-break, the interruption is obviously shorter than with break-before-make.

Another thing that happens after the new IKE_SA and all CHILD_SAs are established is that the old IKE_SA is deleted. For both of these operations, a task is queued that will be activated when everything else is done. The order is currently so that the delete for the old IKE_SA happens before the online certificate verification (or rather concurrently). You can see this in the log:

    16[IKE] activating new tasks
    16[IKE]   activating IKE_REAUTH_COMPLETE task
    16[IKE] activating new tasks
    16[IKE]   activating IKE_VERIFY_PEER_CERT task

The first task only queues an asynchronous delete job for the old IKE_SA, so that's done quickly. The second task performs the online certificate verification in the current thread:

    16[IKE] verifying peer certificate
    16[CFG]   using certificate "<masked>"
    16[CFG]   certificate "<masked>" key: 4096 bit RSA
    16[CFG]   using untrusted intermediate certificate "<masked>"
    16[CFG]   certificate "<masked>" key: 4096 bit RSA
    16[CFG]   using trusted ca certificate "<masked>"
    16[CFG]   certificate "<masked>" key: 4096 bit RSA
    16[CFG]   reached self-signed root ca with a path length of 1
    16[CFG] checking certificate status of "<masked>"
    16[CFG]   requesting ocsp status from 'http://10.197.111.25:80/ejbca/publicweb/status/ocsp' ...

Concurrently, a couple of other threads are handling the delete for the old IKE_SA:

    09[IKE] queueing IKE_DELETE task
    09[IKE] activating new tasks
    09[IKE]   activating IKE_DELETE task
    09[IKE] deleting IKE_SA home[2] between 10.210.101.220[<masked>]...10.197.111.25[<masked>]
    09[IKE] deleting IKE_SA home[2] between 10.210.101.220[<masked>]...10.197.111.25[<masked>]
    09[IKE] IKE_SA home[2] state change: ESTABLISHED => DELETING
    09[IKE] sending DELETE for IKE_SA home[2]
    09[ENC] generating INFORMATIONAL request 31 [ D ]
    09[NET] sending packet: from 10.210.101.220[4500] to 10.197.111.25[4500] (76 bytes)
    12[NET] received packet: from 10.197.111.25[4500] to 10.210.101.220[4500] (76 bytes)
    12[ENC] parsed INFORMATIONAL response 31 [ ]
    12[IKE] IKE_SA deleted
    12[IKE] IKE_SA deleted
    ...
    12[IKE] IKE_SA home[2] state change: DELETING => DESTROYING

While for the new IKE_SA the revocation check is performed and fails, so this IKE_SA is also deleted:

    16[CFG]   ocsp response correctly signed by "<masked>"
    16[CFG] certificate was revoked on Jul 15 14:53:12 UTC 2024, reason: affiliation changed
    16[CFG]   ocsp response is valid: until Jul 15 21:08:26 2024
    16[IKE] peer certificate verification failed, deleting SA
    16[IKE] deleting IKE_SA home[4] between 10.210.101.220[<masked>]...10.197.111.25[<masked>]

We could theoretically reverse the activation order for these two tasks. So the old IKE_SA would only get deleted if the certificate verification was successful. However, since the replacement SAs are fully established at this point, the delete for the old IKE_SA is definitely warranted (traffic can flow via new CHILD_SAs).

The question is basically whether you'd actually still want to keep the old SAs established if you don't trust the peer anymore due to the failed online certificate verification. I'd argue the current order might be safer.
